src/scripts/main.js

```diff
@@ -1,3 +1,388 @@
 'use strict';
 
-// write your code here
+// Game state variables
+let gameGrid = [];
+let score = 0;
+let gameOver = false;
+let win = false;
+
+// DOM elements
+const button = document.querySelector('button.button');
+const scoreDisplay = document.querySelector('.game-score');
+const messageStart = document.querySelector('.message-start');
+const messageWin = document.querySelector('.message-win');
+const messageLose = document.querySelector('.message-lose');
+
+// Start/Restart game
+function startGame() {
+  // Reset game state
+  gameGrid = [];
+  score = 0;
+  gameOver = false;
+  win = false;
+
+  // Generate initial cells
+  generateInitialCells();
+
+  // Set up event listener for keyboard input
+  document.addEventListener('keydown', handleKeyDown);
+
+  // Update UI
+  updateUI();
+}
+
+function resetGame() {
+  button.classList.replace('restart', 'start');
+  button.innerText = 'Start';
+  messageStart.hidden = false;
+
+  gameGrid = [];
+  score = 0;
+  gameOver = true;
+  win = false;
+
+  for (let i = 0; i < 4; i++) {
+    gameGrid[i] = [];
+
+    for (let j = 0; j < 4; j++) {
+      gameGrid[i][j] = null;
+    }
+  }
+
+  updateUI();
+
+  messageStart.classList.remove('hidden');
+  messageLose.classList.add('hidden');
+  messageWin.classList.add('hidden');
+}
+
+// Generate initial cells
+function generateInitialCells() {
+  for (let i = 0; i < 4; i++) {
+    gameGrid[i] = [];
+
+    for (let j = 0; j < 4; j++) {
+      gameGrid[i][j] = null;
+    }
+  }
+
+  generateNewCell();
+  generateNewCell();
+}
+
+// Generate new cell (2 or 4)
+function generateNewCell() {
+  const emptyCells = getEmptyCells();
+
+  if (emptyCells.length === 0) {
+    // No empty cells, game over
+    gameOver = true;
+
+    return;
+  }
+
+  const randomIndex = Math.floor(Math.random() * emptyCells.length);
+  const randomCell = emptyCells[randomIndex];
+
+  // Randomly choose 2 or 4 with a 10% probability for 4
+  const newValue = Math.random() < 0.9 ? 2 : 4;
+
+  // Update game grid and cell values
+  gameGrid[randomCell.row][randomCell.col] = newValue;
+}
+
+// Handle keyboard input
+function handleKeyDown(e) {
+  if (e.key === 'ArrowUp') {
+    moveCellsUp();
+  } else if (e.key === 'ArrowDown') {
+    moveCellsDown();
+  } else if (e.key === 'ArrowLeft') {
+    moveCellsLeft();
+  } else if (e.key === 'ArrowRight') {
+    moveCellsRight();
+  }
+
+  // Generate new cell after each valid move
```

Comment:
After handling keyboard input, a new cell should only be generated if the move was valid and resulted in a change in the game grid.